pkg/bloomcompactor/chunkcompactor.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return nil, err
+	}
+	builder := PersistentBlockBuilder{
+		builder:  b,
+		localDst: localDst,
+	}
+	return &builder, nil
+}
+
+func (p *PersistentBlockBuilder) BuildFrom(itr v1.Iterator[v1.SeriesWithBloom]) (uint32, error) {
+	return p.builder.BuildFrom(itr)
+}
+
+func (p *PersistentBlockBuilder) Data() (io.ReadCloser, error) {
+	blockFile, err := os.Open(filepath.Join(p.localDst, v1.BloomFileName))
+	if err != nil {
+		return nil, err
+	}
+	return blockFile, nil
+}
+
+func makeChunkRefs(chksMetas []tsdbindex.ChunkMeta, tenant string, fp model.Fingerprint) []chunk.Chunk {
+	chunkRefs := make([]chunk.Chunk, 0, len(chksMetas))
+	for _, chk := range chksMetas {
+		chunkRefs = append(chunkRefs, chunk.Chunk{
+			ChunkRef: logproto.ChunkRef{
+				Fingerprint: uint64(fp),
+				UserID:      tenant,
+				From:        chk.From(),
+				Through:     chk.Through(),
+				Checksum:    chk.Checksum,
+			},
+		})
+	}
+
+	return chunkRefs
+}
+
+func buildBloomFromSeries(seriesMeta seriesMeta, fpRate float64, tokenizer compactorTokenizer, chunks []chunk.Chunk) v1.SeriesWithBloom {
+	// Create a bloom for this series
+	bloomForChks := v1.SeriesWithBloom{
+		Series: &v1.Series{
+			Fingerprint: seriesMeta.seriesFP,
+		},
+		Bloom: &v1.Bloom{
+			ScalableBloomFilter: *filter.NewDefaultScalableBloomFilter(fpRate),
+		},
+	}
+
+	// Tokenize data into n-grams
+	tokenizer.PopulateSeriesWithBloom(&bloomForChks, chunks)
+	return bloomForChks
+}
+
+// TODO Test this when bloom block size check is implemented
+func buildBlockFromBloom(
+	ctx context.Context,
+	logger log.Logger,
+	builder blockBuilder,
+	blooms []v1.SeriesWithBloom,
+	job Job,
+) (bloomshipper.Block, error) {
+	// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
+	if err := ctx.Err(); err != nil {
+		return bloomshipper.Block{}, err
+	}
+
+	checksum, err := builder.BuildFrom(v1.NewSliceIter(blooms))
+	if err != nil {
+		level.Error(logger).Log("writing bloom", err)
+		return bloomshipper.Block{}, err
+	}
+
+	data, err := builder.Data()
+	if err != nil {
+		level.Error(logger).Log("reading bloom data", err)
+		return bloomshipper.Block{}, err
+	}
+
+	block := bloomshipper.Block{
+		BlockRef: bloomshipper.BlockRef{
+			Ref: bloomshipper.Ref{
+				TenantID:       job.tenantID,
+				TableName:      job.tableName,
+				MinFingerprint: uint64(job.minFp),
+				MaxFingerprint: uint64(job.maxFp),
+				StartTimestamp: int64(job.from),
+				EndTimestamp:   int64(job.through),
+				Checksum:       checksum,
+			},
+			IndexPath: job.indexPath,
+		},
+		Data: data,
+	}
+
+	return block, nil
+}
+
+func createLocalDirName(workingDir string, job Job) string {
+	dir := fmt.Sprintf("bloomBlock-%s-%s-%s-%s-%s-%s", job.tenantID, job.tenantID, job.minFp, job.maxFp, job.from, job.through)
+	return filepath.Join(workingDir, dir)
+}
+
+// Compacts given list of chunks, uploads them to storage and returns a list of bloomBlocks
+func CompactNewChunks(
```

Comment:
IMO this function should not be exported, since it's meant for package internal use only

```suggestion
func compactNewChunks(
```